### DISCUSSION TITLE
Vercel Dev doesn't persist Node.js endpoint context anymore

### DISCUSSION THREAD

**Discussion by sara-wvs** [vercel/vercel#4928] [Help] [answered]
Variables declared in modules (outside of module.exports) persist when using vercel v19.0.1, but not when using v19.1.2 or v19.2.0.

### Steps to reproduce

0. Create a new folder (any name) and cd into it
1. Create new package.json using `npm init`. (Use whatever name and license, use index.js as entry point)
2. Create a folder called “api” in the project root 
3. Create an index.js file in the “api” folder with the following content:

```
let counter = 0;

module.exports = (req, res) => {
  counter++;

  res.json({
    body: { counter },
  });
};
```
4. In package.json, add "vercel" as a script and vercel ~19.0.0 as a dev dependency:
```
{
  "name": "repro-for-vercel",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "vercel": "vercel"
  },
  "author": "",
  "license": "MIT",
  "devDependencies": {
    "vercel": "~19.0.0"
  }
}

```
5. Run `npm i`
6. Run `npm run vercel` to set up
7. Run `npm run vercel dev`
8. Hit http://localhost:3000/api repeatedly

Expected result: the body of the response contains a `counter` field that increments every time you hit the endpoint
Actual result: as expected ✅ 

9. Change the version of vercel in package.json to ~19.1.0 or ~19.2.0 (same result for both)
```
 "devDependencies": {
    "vercel": "~19.1.0"
  }
```
10. Run `npm i`
11. Run `npm run vercel dev`
12. Hit http://localhost:3000/api repeatedly

Expected result: the body of the response contains a `counter` field that increments every time you hit the endpoint
Actual result: the body of the response contains a `counter` field that is always 1, regardless how many times you hit the endpoint 😞 

The real-life use for this is caching calls to an external API; my cache is always empty when running `vercel dev` but caching functionality seems to be working fine in production.

Is this a bug? Am I doing something wrong?

**Comment by styfle** [comment]
Hi @sara-wvs 

Thanks for the detailed repro steps!

You're right, `vc dev` acts a little differently than an actual deployment because the function is 100% stateless in dev and it will say warm in for some time in prod (eventually it will also reset internal state after inactivity).

This behavior changed in Vercel 19.1.0 for the sake of performance but I see that this could cause problems since you cannot cache any state using global variables with `vc dev`.

I created an Issue (ticket 7747) for tracking this bug, thanks!

cc @TooTallNate

**Comment by leo** [comment, marked as answer]
Thank you for sharing this issue with us!

As explained by @styfle above, this is actually by design, as `vercel dev` re-invokes your Serverless Function fresh every time a request comes in. I unfortunately don't see a way for us to solve this other than redesigning the development architecture entirely.

I understand that this might cause problems if you're relying on storing information between invocations, but I don't think there's a guarantee that AWS Lambdas will remain the same between invocations either. It just happens to usually be the case, as they're optimizing for performance. 

As you can see in the [best practises section](https://docs.aws.amazon.com/lambda/latest/dg/best-practices.html) in the AWS documentation, it's indeed recommended to initialize things like SDKs outside the Serverless Function, such that the instance can be re-used between requests.

However, that doesn't mean your code should **rely** on it. From my perspective, it's impossible to rely on it.
